### ISSUE TITLE
[5.2] failure in `test_writes_to_recent_previous_cdc_generations`: `cdc::metadata::get_stream: could not find any CDC stream`

### ISSUE THREAD

**Issue by kbr-scylla** [scylladb/scylladb#17749]
```
=================================== FAILURES ===================================
________________ test_writes_to_recent_previous_cdc_generations ________________

request = <FixtureRequest for <Function test_writes_to_recent_previous_cdc_generations>>
manager = <test.pylib.manager_client.ManagerClient object at 0x7fe0017a1050>

    @pytest.mark.asyncio
    async def test_writes_to_recent_previous_cdc_generations(request, manager: ManagerClient):
        """
        Test that writes to previous CDC generations succeed if the timestamp of the generation being written to
        is greater than (now - generation_leeway)
    
        The test starts 3 nodes, and thus, it creates 3 generations. Let's denote their timestamps by ts1, ts2, ts3,
        where ts1 is the timestamp of the first (and oldest) generation, and so on. We create a scenario where
        now - generation_leeway < ts1 < ts2 < ts3 < now.
        We check that writes with a timestamp from the interval [ts1, now] succeed.
    
        The value of generation_leeway is 5 s. So, the condition
        now - generation_leeway < ts1
        might be false if the test runs too slowly. To avoid this, we increase generation_leeway to 5 min
        through the error injection.
        """
        logger.info("Bootstrapping nodes")
        servers = [await manager.server_add() for _ in range(3)]
    
        for srv in servers:
            await manager.api.enable_injection(srv.ip_addr, 'increase_cdc_generation_leeway', False)
    
        cql = manager.cql
    
        logger.info("Waiting for driver")
        await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
    
        gen_timestamps = await wait_for_publishing_generations(cql, servers)
    
        logger.info("Ceating a test table")
        await cql.run_async("CREATE KEYSPACE test WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1}")
        await cql.run_async("CREATE TABLE test.test (pk int PRIMARY KEY, c int) WITH cdc = {'enabled': true}")
    
        async def do_write(timestamp: int):
            await cql.run_async(f"INSERT INTO test.test (pk, c) VALUES (0, 0) USING TIMESTAMP {timestamp}")
    
        logger.info("Executing writes to all generations")
        for ts in gen_timestamps:
            # Writes to all three generations should succeed.
            await do_write(ts + 1)
            await do_write(ts)
            # A write with a timestamp smaller than the oldest generation's one should fail.
            if ts == min(gen_timestamps):
                with pytest.raises(NoHostAvailable):
                    await do_write(ts - 1)
            else:
>               await do_write(ts - 1)

test/topology_custom/test_writes_to_previous_cdc_generations.py:90: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

timestamp = 1710185266383999

    async def do_write(timestamp: int):
>       await cql.run_async(f"INSERT INTO test.test (pk, c) VALUES (0, 0) USING TIMESTAMP {timestamp}")
E       cassandra.cluster.NoHostAvailable: ('Unable to complete the operation against any hosts', {<Host: 127.104.119.29:9042 datacenter1>: <Error from server: code=0000 [Server error] message="cdc::metadata::get_stream: could not find any CDC stream for timestamp 2024/03/11 19:27:46. Are we in the middle of a cluster upgrade?">, <Host: 127.104.119.14:9042 datacenter1>: <Error from server: code=0000 [Server error] message="cdc::metadata::get_stream: could not find any CDC stream for timestamp 2024/03/11 19:27:46. Are we in the middle of a cluster upgrade?">, <Host: 127.104.119.10:9042 datacenter1>: <Error from server: code=0000 [Server error] message="cdc::metadata::get_stream: could not find any CDC stream for timestamp 2024/03/11 19:27:46. Are we in the middle of a cluster upgrade?">})
```

In 5.2 the test is running in gossip-mode (there is no consistent-topology-changes mode), so generations are distributed with Everywhere table + gossip. Maybe that's the reason and the test needs to wait until all nodes learn about a generation before writing?

[scylla-373.log](https://github.com/scylladb/scylladb/files/14569783/scylla-373.log)
[scylla-377.log](https://github.com/scylladb/scylladb/files/14569785/scylla-377.log)
[scylla-381.log](https://github.com/scylladb/scylladb/files/14569786/scylla-381.log)
[topology_custom.test_writes_to_previous_cdc_generations.1.log](https://github.com/scylladb/scylladb/files/14569787/topology_custom.test_writes_to_previous_cdc_generations.1.log)

**Comment by kbr-scylla** [comment]
Node scylla-377 learned about only one generation, scylla-377 only about two.

Right, in gossip mode a booting node learns only about the last CDC generation, not the latest few (operating ones) like in raft-topology mode. The test doesn't really make sense in gossip mode...

**Comment by kbr-scylla** [comment]
Alternatively we could adapt it so the writes are coordinated always by the first node which learned about all generations

**Comment by kbr-scylla** [comment]
We likely have the same problem in 5.4, so we need to fix there too.

**Comment by kbr-scylla** [comment]
Ah, but wait... according to the error message, the driver tried all hosts, and failed on all of them. It tried the first node too (scylla-373, IP 127.104.119.14), which knew of all generations...
`<Host: 127.104.119.14:9042 datacenter1>: <Error from server: code=0000 [Server error] message="cdc::metadata::get_stream: could not find any CDC stream for timestamp 2024/03/11 19:27:46. Are we in the middle of a cluster upgrade?">`

**Comment by kbr-scylla** [comment]
From scylla-373:
```
INFO  2024-03-11 21:27:42,587 [shard 0] cdc - Latest generation seen during startup: (2024/03/11 19:27:42, 604737ae-9886-45cd-8bfe-8f980ed603b9)
INFO  2024-03-11 21:27:42,588 [shard 0] cdc - Starting to use generation (2024/03/11 19:27:42, 604737ae-9886-45cd-8bfe-8f980ed603b9)
INFO  2024-03-11 21:27:42,589 [shard 0] cdc - CDC description table successfully updated with generation (2024/03/11 19:27:42, 604737ae-9886-45cd-8bfe-8f980ed603b9).
```

**Comment by kbr-scylla** [comment]
The node probably garbage-collected the generation too fast, I broke the test during backport:
```
    Backport note: replaced `servers_add` with `server_add` loop in tests
    replaced `error_injections_at_startup` (not implemented in 5.2) with
    `enable_injection` post-boot
```

It's important to increase `generation_leeway` during node boot so it prevents GC from happening. I did this instead after all servers booted
```python3
        for srv in servers:
            await manager.api.enable_injection(srv.ip_addr, 'increase_cdc_generation_leeway', False)
```
which is too late.
